[N2PN DOCUMENTATION] New: General routing and broadcast routing

Ne02ptzero opened the pull request:
```
 _____________________________________
/ It were not best that we should all \
| think alike; it is difference of    |
| opinion that makes horse-races.     |
|                                     |
| -- Mark Twain, "Pudd'nhead Wilson's |
\ Calendar"                           /
 -------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

```

Ne02ptzero committed:
.

Ne02ptzero committed:
.

Ne02ptzero committed:
.

Ne02ptzero committed:
.

Ne02ptzero committed:
Merge branch 'work/louis/routing' of github.com:Meshnet/Documentation into work/louis/routing

Ne02ptzero committed:
.

Ne02ptzero committed:
.

Ne02ptzero committed:
.

pgerhardt commented:
Propose adding a section in the header on "1.4 Motivation" to discuss reasons for needing to develop a custom protocol for this application.

mdombrowski1997 committed:
minor grammar and spelling edits

Ne02ptzero committed:
Merge pull request #4 from mdombrowski1997/work/louis/routing

minor grammar and spelling edits

0x5c committed:
- Spelling, grammar, and typos correction.
- Some wording changes
- Some formating correction

0x5c committed:
White-space trimming

0x5c committed:
Wording changes

Ne02ptzero committed:
Merge pull request #5 from meshnet/work/0x5c/routing

All kinds of corrections

Ne02ptzero committed:
.

0x5c commented:
So, comments time!

**Section 4.1.1** has provisions for _multi-user_ access to the nodes, notably the use of `User ID`s. Do we really want to have multi-user per node (Added complexity for both the D2N and N2N side), 
or do we want to remove it, or leave it in the RFC _"for later use"_.
**Section 4.1.1** also describes the use of the `Destination User ID` field as a `Channel` field. This limits the channel ID (the channel's name) to two bytes, which is a bit problematic for channel naming.

In **section 4.1.3**, it should be clear that:
- A node shall never ACK a _channel chat_ message
- _Channel chat_ msgs are broadcasts by nature (should be retransmitted by everyone). _Perhaps moving the channel chat to `BROADCAST`, and create a `Category` field (either `ALERT`, `weather`, or [Channel Name])? To Be Discussed_

In **section 4.2.1**, the `Issued Timestamp` field is described as the date the `BROADCAST`'s validity period begins (with possibility of it being in the future). It should be, however, the date the `BROADCAST` *was* issued by the device (or the original _Issued:_ date in the case of _gov't/weather agency_ alert relaying).

In **section 5**, paragraphs are massive and daunting to read. They should be segmented in smaller paragraphs (just add line-breaks).

In **section 5.2.1**, the use of the _exit node_ terminology might be confusing with the concept of nodes relaying information to/from the internet or other network, commonly called _exit nodes_.


I also recommend the addition of the obligation of every node to drop non-compliant packets.

Ne02ptzero committed:
Merge branch 'master' into work/louis/routing

 _________________________________________
/ It is right that he too should have his \
| little chronicle, his memories, his     |
| reason, and be able to recognize the    |
| good in the bad, the bad in the worst,  |
| and so grow gently old all down the     |
| unchanging days and die one day like    |
| any other day, only shorter. -- Samuel  |
\ Beckett, "Malone Dies"                  /
 -----------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

Signed-off-by: Ne02ptzero <louis@ne02ptzero.me>

Ne02ptzero commented:
> **Section 4.1.1** has provisions for multi-user access to the nodes, notably the use of User IDs. Do we really want to have multi-user per node (Added complexity for both the D2N and N2N side),
or do we want to remove it, or leave it in the RFC "for later use".

I think we should keep it, and implement it. It is a pretty simple feature, easily added from the beginning.

> **Section 4.1.1** also describes the use of the Destination User ID field as a Channel field. This limits the channel ID (the channel's name) to two bytes, which is a bit problematic for channel naming.

Channel ID can be an integer identifier, and each client can keep an internal name table. The information can be covered with a `JOIN CHANNEL` packet or `NEW CHANNEL`, something like that.

> **In section 4.1.3**, it should be clear that:

>    A node shall never ACK a channel chat message
>    Channel chat msgs are broadcasts by nature (should be retransmitted by everyone). Perhaps moving the channel chat to BROADCAST, and create a Category field (either ALERT, weather, or [Channel Name])? To Be Discussed

Why channels would be Broacasts? Channel can be between X peoples, not entire network. As for the ACKs of channel message, I think a client receiving a channel chat should ack reception to the original sender. Chat message are to important to not be ACK-ed.

> In **section 4.2.1**, the Issued Timestamp field is described as the date the BROADCAST's validity period begins (with possibility of it being in the future). It should be, however, the date the BROADCAST was issued by the device (or the original Issued: date in the case of gov't/weather agency alert relaying).

Why would we need the issued date? Do we need to relay other broadcasts than ours?

> In **section 5**, paragraphs are massive and daunting to read. They should be segmented in smaller paragraphs (just add line-breaks).

Yup, i'll fix it.

> In **section 5.2.1**, the use of the exit node terminology might be confusing with the concept of nodes relaying information to/from the internet or other network, commonly called exit nodes.

Agreed too, i'll modify it.

> I also recommend the addition of the obligation of every node to drop non-compliant packets.

It was implicit in my head, i'll write it.

0x5c commented:
So the big misunderstanding: _channels_.
Channels are like forum "topics"; they are

- Public
- Listed on a page
- Created when a first post is posted [to it]

Given these points, a given channel need to:

- Exist when at-least one post is part of it
- Have it's name defined by the first post (and not a two-byte number)
- Be relayed to the whole network to ensure availability

This means that using the `Destination User ID` field for the channel ID isn't going to work. And since they have to be relayed by the whole network, they essentially are broadcasts.

Ne02ptzero committed:
[DOCUMENTATION] NEW: Format, now working with groff format as input.

Signed-off-by: Ne02ptzero <louis@ne02ptzero.me>

Ne02ptzero committed:
[DOCUMENTATION] FIX: 0x5c remarks

Signed-off-by: Ne02ptzero <louis@ne02ptzero.me>

Ne02ptzero committed:
[DOCUMENTATION] NEW: Slight rework of the chat SERVICE, for better
channels

Signed-off-by: Ne02ptzero <louis@ne02ptzero.me>

Ne02ptzero committed:
[DOCUMENTATION] FIX: Now broadcasting channels

Signed-off-by: Ne02ptzero <louis@ne02ptzero.me>

0x5c committed:
[N2NP] Makefile: corrected groff command line

0x5c committed:
.gitignore: cleanup

Ne02ptzero committed:
Merge pull request #9 from meshnet/0x5c/n2np

Minor fixes

0x5c committed:
[REPOSITORY] Directory structure changes
Why? Because the Documentation repo is for all the project documentation.

0x5c committed:
.gitignore[all]: added some EOF newlines for git to shut up

0x5c commented:
The `groff` document does not produce pages. (One long document with no headers/footers)

Ne02ptzero committed:
Merge pull request #10 from meshnet/0x5c/n2np

[REPOSITORY] Directory scruct changes

Ne02ptzero committed:
[DOCUMENTATION] FIX: Paths inside Makefiles and references database

 ______________________________________
/ You will have good luck and overcome \
\ many hardships.                      /
 --------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

Signed-off-by: Ne02ptzero <louis@ne02ptzero.me>

Ne02ptzero committed:
[DOCUMENTATION] FIX: Wrong groff command line option

Signed-off-by: Ne02ptzero <louis@ne02ptzero.me>

0x5c committed:
[PROTOCOLS] Added troubleshooting info

0x5c committed:
[PROTOCOLS] Changed info file's name

0x5c committed:
Slight fixes

0x5c committed:
yup, again

0x5c committed:
[PROTOCOLS] README.md: update

0x5c committed:
[PROTOCOLS] Added .gitattributes
for correct language recognition of .nroff files.

0x5c committed:
fic incomplete commit

Ne02ptzero committed:
Merge pull request #13 from meshnet/0x5c/n2np

[PROTOCOLS] troubleshooting info

0x5c reviewed (approved):
LGTM

JetseVerschuren commented on protocols/README.md:
```diff
@@ -0,0 +1,7 @@
+
+1. [OPTIONAL/RECOMMENDED] Use a true linux/unix install. *GNUwin32* and similar software collections may not behave correctly.
+
+2. Make sure that `groff` is correctly installed. (`apt install groff` or similar) *As per issue #11*
```
I would add a bit of explanation on Ubuntu half installs of groff.

0x5c replied:
@JetseVerschuren I'll do that after the merge (should happen today or tomorrow)
